src/arch/ssse3.rs

```diff
@@ -0,0 +1,262 @@
+    data[2] = _mm_unpacklo_epi64(d0123lh, d4567lh);
+    data[3] = _mm_unpackhi_epi64(d0123lh, d4567lh);
+    data[4] = _mm_unpacklo_epi64(d0123hl, d4567hl);
+    data[5] = _mm_unpackhi_epi64(d0123hl, d4567hl);
+    data[6] = _mm_unpacklo_epi64(d0123hh, d4567hh);
+    data[7] = _mm_unpackhi_epi64(d0123hh, d4567hh);
+}
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+#[target_feature(enable = "ssse3")]
+pub unsafe fn dequantize_and_idct_block_8x8(
+    coefficients: &[i16],
+    quantization_table: &[u16; 64],
+    output_linestride: usize,
+    output: &mut [u8],
+) {
+    assert!(coefficients.len() >= 64);
+    // The loop below will write to positions [output_linestride * i, output_linestride * i + 8)
+    // for 0<=i<8. Thus, the last accessed position is at an offset of output_linestrade * 7 + 7,
+    // and if that position is in-bounds, so are all other accesses.
+    assert!(
+        output.len()
+            > output_linestride
+                .checked_mul(7)
+                .unwrap()
+                .checked_add(7)
+                .unwrap()
+    );
+
+    #[cfg(target_arch = "x86")]
+    use std::arch::x86::*;
+    #[cfg(target_arch = "x86_64")]
+    use std::arch::x86_64::*;
+
+    const SHIFT: i32 = 3;
+
+    let mut data = [_mm_setzero_si128(); 8];
+    for i in 0..8 {
+        data[i] = _mm_slli_epi16(
+            _mm_mullo_epi16(
+                _mm_loadu_si128(coefficients.as_ptr().wrapping_add(i * 8) as *const _),
+                _mm_loadu_si128(quantization_table.as_ptr().wrapping_add(i * 8) as *const _),
+            ),
+            SHIFT,
+        );
+    }
+
+    idct8(&mut data);
+    transpose8(&mut data);
+    idct8(&mut data);
+    transpose8(&mut data);
+
+    for i in 0..8 {
+        let mut buf = [0u8; 16];
+        _mm_storeu_si128(
+            buf.as_mut_ptr() as *mut _,
+            _mm_packus_epi16(
+                _mm_srai_epi16(
+                    _mm_adds_epi16(data[i], _mm_set1_epi16(257 << (SHIFT + 2))),
+                    SHIFT + 3,
+                ),
+                _mm_setzero_si128(),
+            ),
+        );
+        std::ptr::copy_nonoverlapping::<u8>(
+            buf.as_ptr(),
+            output.as_mut_ptr().wrapping_add(output_linestride * i) as *mut _,
+            8,
+        );
+    }
+}
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+#[target_feature(enable = "ssse3")]
+pub unsafe fn color_convert_line_ycbcr(y: &[u8], cb: &[u8], cr: &[u8], output: &mut [u8]) -> usize {
+    assert!(output.len() % 3 == 0);
+    let num = output.len() / 3;
+    assert!(num <= y.len());
+    assert!(num <= cb.len());
+    assert!(num <= cr.len());
+    // _mm_loadu_si64 seems to be confused between stable and beta on whether it loads on the low
+    // or the upper half of the vector. To circumvent the issue, we use _mm_loadu_si128, and skip
+    // one vector to avoid loads from outside accessible memory.
+    let num_vecs = (num / 8).saturating_sub(1);
```

Comment:
Sounds nice in theory but what happens to the last vector now? There doesn't appear to be a fixup at the end. Also, the path to properly fixing this issue is unclear to me. As far as I understand the correct instruction _could_ be used starting with Rust `1.58`. However, doing so would clearly lead to incorrect code in Rust `<1.57`. Clearly, this version number should be added in the comment here somewhere so that will be picked up for the msrv description if we decide to go back to the 'proper' intrinsic.

Reply:
That's not an issue - leftover pixels are handled by the non-SIMD code (and they'd need to be either way for images that are not multiples of 8/16 pixels). So this change only increases that number a little bit. I'll update the comment to explicitly mention the Rust version.

Reply:
Ah I see, the change has also implicitly influence the number of handled pixels that's being returned. A smarter implementation than my quick review of the PR :sweat_smile: Very good.